### PR TITLE
User login

### PR DESCRIPTION
**Pull request by raj-al-ghul**
part 2 of #12

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -2,6 +2,11 @@ class UsersController < ApplicationController
   rescue_from ActiveRecord::RecordNotFound, with: :redirect_to_index
 
   def new
+    if logged_in?
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Is there a test for this? What about the create action?

**Comment by bboe** [comment]
Just as a heads up @grahamotte will take over as reviewer for the remainder of part 2. I'll still review part 3.

File: test/controllers/sessions_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    user = new_user
+    post :create, params: { session: { email: user.email, password: user.password } }
+    assert_redirected_to images_path
+    assert_equal 'You have succesfully logged in!', flash[:success]
+  end
+
+  test 'destroy should log out a user' do
+    user = new_user
+    @controller.log_in user
+    delete :destroy, params: { id: user.id }
```
[the remark this answers is not in the document]
Comment (raj-al-ghul) [reply test/controllers/sessions_controller_test.rb]: not needed (actually wasn't used), just the default resource path needed it

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,8 @@
 Rails.application.routes.draw do
+  resources :sessions, only: [:new, :create]
+
+  delete 'logout', to: 'sessions#destroy'
```
Comment (grahamotte) [inline config/routes.rb]: This can be a part of `sessions`s resources

File: test/controllers/sessions_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    user = create_user
+    @controller.log_in user
+    delete :destroy
+    assert_redirected_to root_path
+    assert_equal 'You have succesfully logged out!', flash[:success]
+  end
+
+  private
+
+  def create_user(name: 'ned', email: 'ned@gmail.com', password: 'abcdefgh', password_confirmation: 'abcdefgh')
```
Comment (grahamotte) [inline test/controllers/sessions_controller_test.rb]: Consider pulling these common tasks out into their own helpers

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -19,4 +20,10 @@ def create
   def create_params
     params.require(:user).permit(:name, :email, :password, :password_confirmation)
   end
+
+  def duplicate_login
```
Comment (grahamotte) [inline app/controllers/users_controller.rb]: Consider changing this method name to reflect the action it takes. Something like `redirect_if_already_logged_in`

Comment (grahamotte) [reply app/controllers/users_controller.rb]: This also is duplicated on the `SessionsController`. Can we DRY it?

File: test/test_helpers/user_test_helper.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module UserTestHelper
+  def user_params(name: 'ned', email: 'ned@gmail.com', password: 'abcdefgh', password_confirmation: 'abcdefgh')
+    { user: { name: name, email: email, password: password, password_confirmation: password_confirmation } }
+  end
+
+  def create_user(name: 'ned', email: 'ned@gmail.com', password: 'abcdefgh', password_confirmation: 'abcdefgh')
```
Comment (grahamotte) [inline test/test_helpers/user_test_helper.rb]: what happens if you call `create_user ` twice?

File: test/controllers/users_controller_test.rb [hunk 7]
```diff
@@ -1,12 +1,38 @@
 require 'test_helper'
+require 'test_helpers/user_test_helper'
```
Comment (grahamotte) [inline test/controllers/users_controller_test.rb]: Can we include this in the test helper so we dont have to require it every time?

File: test/controllers/sessions_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,53 @@
+require 'test_helper'
+require 'test_helpers/user_test_helper'
+
+class SessionsControllerTest < ActionController::TestCase
+  include UserTestHelper
+  tests SessionsController
```
Comment (grahamotte) [inline test/controllers/sessions_controller_test.rb]: Is this needed?

File: test/controllers/users_controller_test.rb (hunk 7 above)
Comment (raj-al-ghul) [reply test/controllers/users_controller_test.rb]: i thought the pro of having in its own file was so we don't include helpers in all the tests that don't make use of it

File: test/controllers/sessions_controller_test.rb (hunk 8 above)
Comment (raj-al-ghul) [reply test/controllers/sessions_controller_test.rb]: i believe so (unless i include the helper in test helper)

File: test/controllers/users_controller_test.rb (hunk 7 above)
Comment (grahamotte) [reply test/controllers/users_controller_test.rb]: Yeah it's kind of what you prefer. I would say though for these super common kinds of things then the overhead of always including it is worth it.

File: test/controllers/sessions_controller_test.rb (hunk 8 above)
Comment (grahamotte) [reply test/controllers/sessions_controller_test.rb]: Ah I see you're using it to get the helper

File: test/test_helpers/user_test_helper.rb (hunk 6 above)
Comment (raj-al-ghul) [reply test/test_helpers/user_test_helper.rb]: just calling it twice will fail with email already taken, i changed it to named params so it'd be easy to call it again with a different email

Comment (grahamotte) [reply test/test_helpers/user_test_helper.rb]: Consider adding a counter to the unique fields. As it is you have to know to change the params after the first call which leads to unexpected behavior

File: test/test_helper.rb [hunk 9]
```diff
@@ -1,12 +1,19 @@
...
 module ActiveSupport
   class TestCase
+    include UserTestHelper
+
     # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
     fixtures :all
 
     # Add more helper methods to be used by all tests here...
+    # Returns true if a test user is logged in.
+    def logged_in?
```
Comment (grahamotte) [inline test/test_helper.rb]: Consider moving this to it's own sessions helper as well

**Pull request merged**